File: app/src/main/java/com/example/bunnix/database/firebase/collections/ChatCollection.kt
package com.example.bunnix.database.firebase.collections

import com.example.bunnix.database.config.FirebaseConfig
import com.example.bunnix.database.models.Chat
import com.example.bunnix.database.models.Message
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.Query
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

object ChatCollection {

    private val chatsCollection = FirebaseConfig.firestore.collection(FirebaseConfig.Collections.CHATS)

    // GET OR CREATE CHAT
    suspend fun getOrCreateChat(customerId: String, vendorId: String): Result<String> {
        return try {
            // Check if chat exists
            val existingChat = chatsCollection
                .whereArrayContains("participants", customerId)
                .get()
                .await()
                .documents
                .firstOrNull { doc ->
                    val participants = doc.get("participants") as? List<*>
                    participants?.contains(vendorId) == true
                }

            if (existingChat != null) {
                return Result.success(existingChat.id)
            }

            // Create new chat
            val chatData = mapOf(
                "participants" to listOf(customerId, vendorId),
                "participantDetails" to emptyMap<String, Any>(),
                "lastMessage" to "",
                "lastMessageTime" to null,
                "lastMessageSender" to "",
                "unreadCount" to mapOf(
                    customerId to 0,
                    vendorId to 0
                ),
                "relatedOrderId" to "",
                "relatedBookingId" to "",
                "createdAt" to Timestamp.now()
            )

            val docRef = chatsCollection.add(chatData).await()
            Result.success(docRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // GET USER CHATS (Real-time)
    fun getUserChats(userId: String): Flow<List<Chat>> = callbackFlow {
        val listener = chatsCollection
            .whereArrayContains("participants", userId)
            .orderBy("lastMessageTime", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val chats = snapshot?.toObjects(Chat::class.java) ?: emptyList()
                trySend(chats)
            }
        awaitClose { listener.remove() }
    }

    // GET MESSAGES (Real-time)
    fun getMessages(chatId: String): Flow<List<Message>> = callbackFlow {
        val listener = chatsCollection
            .document(chatId)
            .collection(FirebaseConfig.Collections.MESSAGES)
            .orderBy("timestamp", Query.Direction.ASCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val messages = snapshot?.toObjects(Message::class.java) ?: emptyList()
                trySend(messages)
            }
        awaitClose { listener.remove() }
    }

    // SEND MESSAGE
    suspend fun sendMessage(
        chatId: String,
        senderId: String,
        senderName: String,
        text: String,
        imageUrl: String = "",
        messageType: String = "text"
    ): Result<String> {
        return try {
            val messageData = mapOf(
                "senderId" to senderId,
                "senderName" to senderName,
                "text" to text,
                "imageUrl" to imageUrl,
                "messageType" to messageType,
                "orderPreview" to emptyMap<String, Any>(),
                "isRead" to false,
                "timestamp" to Timestamp.now()
            )

            // Add message to subcollection
            val messageRef = chatsCollection
                .document(chatId)
                .collection(FirebaseConfig.Collections.MESSAGES)
                .add(messageData)
                .await()

            // Update chat's last message
            updateLastMessage(chatId, senderId, text)

            Result.success(messageRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // SEND MESSAGE WITH ORDER PREVIEW
    suspend fun sendMessageWithOrderPreview(
        chatId: String,
        senderId: String,
        senderName: String,
        text: String,
        orderPreview: Map<String, Any>
    ): Result<String> {
        return try {
            val messageData = mapOf(
                "senderId" to senderId,
                "senderName" to senderName,
                "text" to text,
                "imageUrl" to "",
                "messageType" to "order_link",
                "orderPreview" to orderPreview,
                "isRead" to false,
                "timestamp" to Timestamp.now()
            )

            val messageRef = chatsCollection
                .document(chatId)
                .collection(FirebaseConfig.Collections.MESSAGES)
                .add(messageData)
                .await()

            updateLastMessage(chatId, senderId, text)

            Result.success(messageRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // MARK MESSAGES AS READ
    suspend fun markMessagesAsRead(chatId: String, userId: String): Result<Unit> {
        return try {
            // Reset unread count for user
            val updates = mapOf(
                "unreadCount.$userId" to 0
            )
            chatsCollection.document(chatId).update(updates).await()

            // Mark individual messages as read
            val messagesSnapshot = chatsCollection
                .document(chatId)
                .collection(FirebaseConfig.Collections.MESSAGES)
                .whereEqualTo("isRead", false)
                .get()
                .await()

            val batch = FirebaseConfig.firestore.batch()
            messagesSnapshot.documents.forEach { doc ->
                if (doc.getString("senderId") != userId) {
                    batch.update(doc.reference, "isRead", true)
                }
            }
            batch.commit().await()

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // UPDATE LAST MESSAGE
    private suspend fun updateLastMessage(chatId: String, senderId: String, text: String) {
        try {
            val chatDoc = chatsCollection.document(chatId).get().await()
            val participants = chatDoc.get("participants") as? List<*>
            val otherUserId = participants?.firstOrNull { it != senderId } as? String

            val updates = mutableMapOf<String, Any>(
                "lastMessage" to text,
                "lastMessageTime" to Timestamp.now(),
                "lastMessageSender" to senderId
            )

            // Increment unread count for other user
            if (otherUserId != null) {
                val currentUnread = (chatDoc.get("unreadCount") as? Map<*, *>)?.get(otherUserId) as? Long ?: 0
                updates["unreadCount.$otherUserId"] = currentUnread + 1
            }

            chatsCollection.document(chatId).update(updates).await()
        } catch (e: Exception) {
            // Ignore errors in updating last message
        }
    }
}

File: src/database/firebase/collections/chatCollection.js
import { db, COLLECTIONS } from '../config';
import { 
  collection, 
  doc,
  addDoc, 
  getDoc,
  getDocs,
  updateDoc,
  query, 
  where, 
  orderBy,
  onSnapshot,
  arrayUnion,
  increment,
  serverTimestamp 
} from 'firebase/firestore';

const chatsCollection = collection(db, COLLECTIONS.CHATS);

/**
 * GET OR CREATE CHAT
 */
export const getOrCreateChat = async (customerId, vendorId) => {
  try {
    // Check if chat exists
    const q = query(
      chatsCollection,
      where('participants', 'array-contains', customerId)
    );
    
    const snapshot = await getDocs(q);
    const existingChat = snapshot.docs.find(doc => {
      const participants = doc.data().participants;
      return participants.includes(vendorId);
    });
    
    if (existingChat) {
      return { success: true, chatId: existingChat.id };
    }
    
    // Create new chat
    const chatData = {
      participants: [customerId, vendorId],
      participantDetails: {},
      lastMessage: '',
      lastMessageTime: null,
      lastMessageSender: '',
      unreadCount: {
        [customerId]: 0,
        [vendorId]: 0
      },
      relatedOrderId: '',
      relatedBookingId: '',
      isActive: true,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp()
    };
    
    const docRef = await addDoc(chatsCollection, chatData);
    return { success: true, chatId: docRef.id };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * GET USER CHATS (Real-time)
 */
export const subscribeToUserChats = (userId, callback) => {
  const q = query(
    chatsCollection,
    where('participants', 'array-contains', userId),
    orderBy('lastMessageTime', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const chats = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(chats);
  });
};

/**
 * GET MESSAGES (Real-time)
 */
export const subscribeToMessages = (chatId, callback) => {
  const messagesCollection = collection(db, `${COLLECTIONS.CHATS}/${chatId}/messages`);
  const q = query(messagesCollection, orderBy('timestamp', 'asc'));
  
  return onSnapshot(q, (snapshot) => {
    const messages = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(messages);
  });
};

/**
 * SEND MESSAGE
 */
export const sendMessage = async (chatId, senderId, senderName, text, imageUrl = '', messageType = 'text') => {
  try {
    const messagesCollection = collection(db, `${COLLECTIONS.CHATS}/${chatId}/messages`);
    
    const messageData = {
      senderId,
      senderName,
      text,
      imageUrl,
      messageType,
      orderPreview: {},
      isRead: false,
      timestamp: serverTimestamp(),
      deletedFor: []
    };
    
    await addDoc(messagesCollection, messageData);
    
    // Update chat's last message
    await updateLastMessage(chatId, senderId, text);
    
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * SEND MESSAGE WITH ORDER PREVIEW
 */
export const sendMessageWithOrderPreview = async (chatId, senderId, senderName, text, orderPreview) => {
  try {
    const messagesCollection = collection(db, `${COLLECTIONS.CHATS}/${chatId}/messages`);
    
    const messageData = {
      senderId,
      senderName,
      text,
      imageUrl: '',
      messageType: 'order_link',
      orderPreview,
      isRead: false,
      timestamp: serverTimestamp(),
      deletedFor: []
    };
    
    await addDoc(messagesCollection, messageData);
    await updateLastMessage(chatId, senderId, text);
    
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * UPDATE LAST MESSAGE
 */
const updateLastMessage = async (chatId, senderId, text) => {
  try {
    const chatDoc = await getDoc(doc(chatsCollection, chatId));
    if (!chatDoc.exists()) return;
    
    const participants = chatDoc.data().participants;
    const otherUserId = participants.find(id => id !== senderId);
    
    const currentUnread = chatDoc.data().unreadCount || {};
    
    await updateDoc(doc(chatsCollection, chatId), {
      lastMessage: text,
      lastMessageTime: serverTimestamp(),
      lastMessageSender: senderId,
      [`unreadCount.${otherUserId}`]: (currentUnread[otherUserId] || 0) + 1,
      updatedAt: serverTimestamp()
    });
  } catch (error) {
    console.error('Error updating last message:', error);
  }
};

/**
 * MARK MESSAGES AS READ
 */
export const markMessagesAsRead = async (chatId, userId) => {
  try {
    await updateDoc(doc(chatsCollection, chatId), {
      [`unreadCount.${userId}`]: 0
    });
    
    // Mark individual messages as read
    const messagesCollection = collection(db, `${COLLECTIONS.CHATS}/${chatId}/messages`);
    const q = query(messagesCollection, where('isRead', '==', false));
    const snapshot = await getDocs(q);
    
    const updatePromises = snapshot.docs
      .filter(doc => doc.data().senderId !== userId)
      .map(doc => updateDoc(doc.ref, { isRead: true }));
    
    await Promise.all(updatePromises);
    
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * DELETE MESSAGE FOR USER
 */
export const deleteMessageForUser = async (chatId, messageId, userId) => {
  try {
    const messageRef = doc(db, `${COLLECTIONS.CHATS}/${chatId}/messages`, messageId);
    await updateDoc(messageRef, {
      deletedFor: arrayUnion(userId)
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};